Aplib.Core/Belief/MemoryBelief.cs

```diff
@@ -0,0 +1,104 @@
+        {
+            _memorizedObservations = new(framesToRemember);
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MemoryBelief{TReference, TResource}"/> class.
+        /// </summary>
+        /// <param name="reference">The reference used to generate/update the resource.</param>
+        /// <param name="getResourceFromReference">A function that takes a reference and generates/updates a resource.</param>
+        /// <param name="framesToRemember">The number of frames to remember back.</param>
+        /// <param name="shouldUpdate">A function that sets a condition on when the resource should be updated.</param>
+        public MemoryBelief(TReference reference, Func<TReference, TObservation> getResourceFromReference, int framesToRemember,
+            Func<bool> shouldUpdate)
+            : base(reference, getResourceFromReference, shouldUpdate)
+        {
+            _memorizedObservations = new(framesToRemember);
+        }
+
+        /// <summary>
+        /// Generates/updates the resource.
+        /// Also stores the previous observation in memory.
+        /// </summary>
+        public override void UpdateBelief()
+        {
+            // We use the implicit conversion to TObservation to store the observation
+            _memorizedObservations.Put(this);
+            base.UpdateBelief();
+        }
+
+        /// <summary>
+        /// Gets the most recently memorized resource.
+        /// </summary>
+        /// <returns> The most recent memory of the resource.</returns>
+        public TObservation GetMostRecentMemory() => _memorizedObservations.GetFirst();
+
+        /// <summary>
+        /// Gets the memorized resource at a specific index.
+        /// A higher index means a memory further back in time.
+        /// If the index is out of bounds, returns the closest element that is in bounds.
+        /// </summary>
+        /// <returns> The memory of the resource at the specified index.</returns>
+        public TObservation GetMemoryAt(int index)
+        {
+            if (index < 0) return GetMostRecentMemory();
+            if (index >= _memorizedObservations.Length)
+                return _memorizedObservations[_memorizedObservations.Length - 1];
+            return _memorizedObservations[index];
+        }
+
+        /// <summary>
+        /// Gets all the memorized resources.
+        /// The first element is the newest memory.
+        /// If not all slots are filled, returns a smaller array.
+        /// </summary>
+        /// <returns> An array of all the memorized resources.</returns>
+        public TObservation[] GetAllMemories()
+        {
+            // Keep track of last non-default index in case of empty slots in the middle
+            int lastNonDefaultIndex = -1;
+            for (int i = 0; i < _memorizedObservations.Length; i++)
+            {
+                if (!_memorizedObservations[i]!.Equals(default(TObservation)))
```

Comment:
This could create issues when you store a default value as a memory. Maybe store the number of actual observations somewhere? Possibly slightly adjust the CircularArray data structure, maybe to something like an ExposedQueue (so it would function as a regular Queue where you can only put elements at the start, but can peek inside the Queue at any time)?

Reply:
TODO: fix in later ticket